extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -14,5 +14,85 @@
  * limitations under the License.
  */
 
+import {Messaging} from './messaging.js';
+import {listen} from '../../../src/event-helper';
+import {viewerForDoc} from '../../../src/viewer';
+import {dev} from '../../../src/log';
+
+const TAG = 'amp-viewer-integration';
+
+/**
+ * @fileoverview This is the communication protocol between AMP and the viewer.
+ * This should be included in an AMP html file to communicate with the viewer.
+ */
 export class AmpViewerIntegration {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} win */
+    this.win = win;
+  }
+
+  /**
+   * Initiate the handshake. If handshake confirmed, start listening for
+   * messages.
+   * @return {!Promise}
+   */
+  init() {
+    dev().info(TAG, 'handshake init()');
+    const viewer = viewerForDoc(this.win.document);
+    return this.getHandshakePromise_(viewer)
+      .then(viewerOrigin => {
+        dev().info(TAG, 'listening for messages');
+        const messaging =
+          new Messaging(this.win, this.win.parent, viewerOrigin,
+            (type, payload, awaitResponse) => {
+              return viewer.receiveMessage(
+                type, /** @type {!JSONType} */ (payload), awaitResponse);
+            });
+        viewer.setMessageDeliverer(messaging.sendRequest.bind(messaging),
+          viewerOrigin);
+      });
+  }
+
+  /**
+   * Send a handshake request, and listen for a handshake response to
+   * confirm the handshake.
+   * @param {!../../../src/service/viewer-impl.Viewer} viewer
+   * @return {?Promise}
+   * @private
+   */
+  getHandshakePromise_(viewer) {
+    const win = this.win;
+    return new Promise(resolve => {
+      const unconfirmedViewerOrigin = viewer.getParam('viewerorigin');
```

Comment:
I believe this code should be even higher up - in the `init` method. Returning `null` here would simply propagate `null` `viewerOrigin` deeper and cause problems downstream.